io.katacontainers.config.hypervisor.image pod annotation updates CAA globally

### Describe the bug

I found this problem while running the AWS e2e tests. I don't know yet whether is aws specify or across providers.

I ran the `TestAwsCreatePeerPodContainerWithInvalidAlternateImage` which creates a simple pod with the `io.katacontainers.config.hypervisor.image` annotation, passing a non-existing image ID (`ami-123456`). The pod is expected to fail and, indeed, it fails. However, following I ran the `TestAwsPodWithInitContainer` which is unrelated, and it is still using the wrong image ID, consequently the test fails.

I suspect when you create a pod with `io.katacontainers.config.hypervisor.image` it's change the image ID value globally at CAA's process, so the next pods created will be using that same ID.

### How to reproduce

Run the AWS e2e test suite.

### CoCo version information

latest

### What TEE are you seeing the problem on

None

### Failing command and relevant log output

```shell
Here is an snippet of the execution:


(...)
2025-02-06T14:53:54.0333174Z === NAME  TestAwsCreatePeerPodContainerWithInvalidAlternateImage/PodVMwithAnnotationsInvalidAlternateImage_test
2025-02-06T14:53:54.0334644Z     assessment_runner.go:549: Deleting pod annotations-invalid-alternate-image...
2025-02-06T14:53:59.0423521Z     assessment_runner.go:556: Pod annotations-invalid-alternate-image has been successfully deleted within 60s
2025-02-06T14:53:59.0425993Z --- PASS: TestAwsCreatePeerPodContainerWithInvalidAlternateImage (10.07s)
2025-02-06T14:53:59.0428198Z     --- PASS: TestAwsCreatePeerPodContainerWithInvalidAlternateImage/PodVMwithAnnotationsInvalidAlternateImage_test (10.07s)
2025-02-06T14:53:59.0432577Z         --- PASS: TestAwsCreatePeerPodContainerWithInvalidAlternateImage/PodVMwithAnnotationsInvalidAlternateImage_test/Failed_to_Create_PodVM_with_a_non-existent_image (0.01s)
2025-02-06T14:53:59.0434197Z === RUN   TestAwsPodWithInitContainer
2025-02-06T14:53:59.0434907Z === RUN   TestAwsPodWithInitContainer/PodWithInitContainer_test
2025-02-06T15:03:59.0626807Z     assessment_runner.go:280: timed out waiting for the condition
2025-02-06T15:03:59.0627625Z     assessment_runner.go:283: Waiting for containers in pod: pod-with-init-container are ready
2025-02-06T15:13:59.0767723Z     assessment_runner.go:294: Expected Pod State: Running
2025-02-06T15:13:59.0768745Z     assessment_runner.go:299: Current Pod State: phase: Pending

<snip>

2025-02-06T15:13:59.0814668Z === RUN   TestAwsPodWithInitContainer/PodWithInitContainer_test/Pod_with_init_container
2025-02-06T15:13:59.0869376Z     assessment_runner.go:389: unexpected warning/error event(s): Failed to create pod sandbox: rpc error: code = Unknown desc = CreateContainer failed: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : Creating instance (<nil>) returned error: operation error EC2: RunInstances, https response error StatusCode: 400, RequestID: 06461e3d-19ad-4077-b527-3c141108fd2b, api error InvalidAMIID.NotFound: The image id '[ami-123456]' does not exist: not foundFailed to create pod sandbox: rpc error: code = Unknown desc = CreateContainer failed: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : Creating instance (<nil>) returned error: operation error EC2: RunInstances, https response error StatusCode: 400, RequestID: 2b53078e-725d-4e63-b9e3-7b4c3449db17, api error InvalidAMIID.NotFound: The image id '[ami-123456]' does not exist: not foundFailed to create pod sandbox: rpc error: code = Unknown desc = CreateContainer failed: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : Creating instance (<nil>) returned error: operation error EC2: RunInstances, https response error StatusCode: 400, RequestID: f92fb5bb-ee88-413b-9923-58a6fe73a21d, api error InvalidAMIID.NotFound: The image id '[ami-123456]' does not exist: not foundFailed to create pod sandbox: rpc error: code = Unknown desc = CreateContainer failed: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : Creating instance (<nil>) returned error: operation error EC2: RunInstances, https response error StatusCode: 400, RequestID: 52ed42ba-ad18-48e6-aed0-e1273cce0982, api error InvalidAMIID.NotFound: The image id '[ami-123456]' does not exist: not foundFailed to create pod sandbox: rpc error: code = Unknown desc = CreateContainer failed: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : Creating instance (<nil>) returned error: operation error EC2: RunInstances, https response error StatusCode: 400, RequestID: e4a270a8-2950-40c5-8d4c-aa00d6fda10f, api error InvalidAMIID.NotFound: The image id '[ami-123456]' does not exist: not foundFailed to create pod sandbox: rpc error: code = Unknown desc = CreateContainer failed: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : Creating instance (<nil>) returned error: operation error EC2: RunInstances, https response error StatusCode: 400, RequestID: bceb2611-bd6f-4ad6-855b-5b3dd520075f, api error InvalidAMIID.NotFound: The image id '[ami-123456]' does not exist: not foundFailed to create pod sandbox: rpc error: code = Unknown desc = CreateContainer failed: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : Creating instance (<nil>) returned error: operation error EC2: RunInstances, https response error StatusCode: 400, RequestID: 935170ff-8ab0-4d85-a9b4-1bdf75dcf982, api error InvalidAMIID.NotFound: The image id '[ami-123456]' does not exist: not foundFailed to create pod sandbox: rpc error: code = Unknown desc = CreateContainer failed: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : Creating instance (<nil>) returned error: operation error EC2: RunInstances, https response error StatusCode: 400, RequestID: a9c8b9ef-1663-4dd8-b64a-0291de16ca43, api error InvalidAMIID.NotFound: The image id '[ami-123456]' does not exist: not foundFailed to create pod sandbox: rpc error: code = Unknown desc = CreateContainer failed: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : Creating instance (<nil>) returned error: operation error EC2: RunInstances, https response error StatusCode: 400, RequestID: 8960cef0-f2c0-4810-a355-4ffb7b7ae91c, api error InvalidAMIID.NotFound: The image id '[ami-123456]' does not exist: not found(combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = CreateContainer failed: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : Creating instance (<nil>) returned error: operation error EC2: RunInstances, https response error StatusCode: 400, RequestID: a3ac0ec0-a5da-4136-954f-c6d372a6fbbb, api error InvalidAMIID.NotFound: The image id '[ami-123456]' does not exist: not found
2025-02-06T15:13:59.2929808Z     aws_common.go:70: Podvm name=pod-with-init-container not found

<snip>

2025-02-06T15:13:59.6279133Z     assessment_runner.go:549: Deleting pod pod-with-init-container...
2025-02-06T15:14:04.4625983Z     assessment_runner.go:556: Pod pod-with-init-container has been successfully deleted within 60s
2025-02-06T15:14:04.4626797Z --- FAIL: TestAwsPodWithInitContainer (1205.42s)
2025-02-06T15:14:04.4627554Z     --- FAIL: TestAwsPodWithInitContainer/PodWithInitContainer_test (1205.42s)
2025-02-06T15:14:04.4628421Z         --- FAIL: TestAwsPodWithInitContainer/PodWithInitContainer_test/Pod_with_init_container (0.22s)
2025-02-06T15:14:04.4628928Z FAIL
(...)
```

--- a/src/cloud-providers/aws/provider.go
+++ b/src/cloud-providers/aws/provider.go
@@ -205,15 +205,17 @@ func (p *awsProvider) CreateInstance(ctx context.Context, podName, sandboxID str
 		}
 	} else {
 
+		imageId := p.serviceConfig.ImageId
+
 		if spec.Image != "" {
 			logger.Printf("Choosing %s from annotation as the AWS AMI for the PodVM image", spec.Image)
-			p.serviceConfig.ImageId = spec.Image
+			imageId = spec.Image
 		}
 
 		input = &ec2.RunInstancesInput{
 			MinCount:          aws.Int32(1),
 			MaxCount:          aws.Int32(1),
-			ImageId:           aws.String(p.serviceConfig.ImageId),
+			ImageId:           aws.String(imageId),
 			InstanceType:      types.InstanceType(instanceType),
 			SecurityGroupIds:  p.serviceConfig.SecurityGroupIds,
 			SubnetId:          aws.String(p.serviceConfig.SubnetId),
@@ -337,8 +339,8 @@ func (p *awsProvider) Teardown() error {
 }
 
 func (p *awsProvider) ConfigVerifier() error {
-	ImageId := p.serviceConfig.ImageId
-	if len(ImageId) == 0 {
+	imageId := p.serviceConfig.ImageId
+	if len(imageId) == 0 {
 		return fmt.Errorf("ImageId is empty")
 	}
 	return nil

--- a/src/cloud-providers/azure/provider.go
+++ b/src/cloud-providers/azure/provider.go
@@ -234,12 +234,14 @@ func (p *azureProvider) CreateInstance(ctx context.Context, podName, sandboxID s
 		return nil, err
 	}
 
+	imageId := p.serviceConfig.ImageId
+
 	if spec.Image != "" {
 		logger.Printf("Choosing %s from annotation as the Azure Image for the PodVM image", spec.Image)
-		p.serviceConfig.ImageId = spec.Image
+		imageId = spec.Image
 	}
 
-	vmParameters, err := p.getVMParameters(instanceSize, diskName, cloudConfigData, sshBytes, instanceName, nicName)
+	vmParameters, err := p.getVMParameters(instanceSize, diskName, cloudConfigData, sshBytes, instanceName, nicName, imageId)
 	if err != nil {
 		return nil, err
 	}
@@ -300,8 +302,8 @@ func (p *azureProvider) Teardown() error {
 }
 
 func (p *azureProvider) ConfigVerifier() error {
-	ImageId := p.serviceConfig.ImageId
-	if len(ImageId) == 0 {
+	imageId := p.serviceConfig.ImageId
+	if len(imageId) == 0 {
 		return fmt.Errorf("ImageId is empty")
 	}
 
@@ -371,7 +373,7 @@ func (p *azureProvider) getResourceTags() map[string]*string {
 	return tags
 }
 
-func (p *azureProvider) getVMParameters(instanceSize, diskName, cloudConfig string, sshBytes []byte, instanceName, nicName string) (*armcompute.VirtualMachine, error) {
+func (p *azureProvider) getVMParameters(instanceSize, diskName, cloudConfig string, sshBytes []byte, instanceName, nicName string, imageId string) (*armcompute.VirtualMachine, error) {
 	userDataB64 := base64.StdEncoding.EncodeToString([]byte(cloudConfig))
 
 	// Azure limits the base64 encrypted userData to 64KB.
@@ -406,11 +408,11 @@ func (p *azureProvider) getVMParameters(instanceSize, diskName, cloudConfig stri
 	}
 
 	imgRef := &armcompute.ImageReference{
-		ID: to.Ptr(p.serviceConfig.ImageId),
+		ID: to.Ptr(imageId),
 	}
-	if strings.HasPrefix(p.serviceConfig.ImageId, "/CommunityGalleries/") {
+	if strings.HasPrefix(imageId, "/CommunityGalleries/") {
 		imgRef = &armcompute.ImageReference{
-			CommunityGalleryImageID: to.Ptr(p.serviceConfig.ImageId),
+			CommunityGalleryImageID: to.Ptr(imageId),
 		}
 	}
 

--- a/src/cloud-providers/ibmcloud-powervs/provider.go
+++ b/src/cloud-providers/ibmcloud-powervs/provider.go
@@ -52,14 +52,16 @@ func (p *ibmcloudPowerVSProvider) CreateInstance(ctx context.Context, podName, s
 		return nil, err
 	}
 
+	imageId := p.serviceConfig.ImageId
+
 	if spec.Image != "" {
 		logger.Printf("Choosing %s from annotation as the Power VS image for the PodVM image", spec.Image)
-		p.serviceConfig.ImageId = spec.Image
+		imageId = spec.Image
 	}
 
 	body := &models.PVMInstanceCreate{
 		ServerName:  &instanceName,
-		ImageID:     &p.serviceConfig.ImageId,
+		ImageID:     &imageId,
 		KeyPairName: p.serviceConfig.SSHKey,
 		Networks: []*models.PVMInstanceAddNetwork{
 			{
@@ -148,8 +150,8 @@ func (p *ibmcloudPowerVSProvider) Teardown() error {
 }
 
 func (p *ibmcloudPowerVSProvider) ConfigVerifier() error {
-	ImageId := p.serviceConfig.ImageId
-	if len(ImageId) == 0 {
+	imageId := p.serviceConfig.ImageId
+	if len(imageId) == 0 {
 		return fmt.Errorf("ImageId is empty")
 	}
 	return nil

